### PR TITLE
Fix physgrid issues with PE layout and aqua restarts

### PR DESCRIPTION
**Pull request by whannah1**
This fixes an issues where aqua planet cases were crashing on restart due to not correctly setting the topography to zero. Instead the realistic topography was added on restart, which caused the run to crash. This PR resolves this issue by explicitly resetting the topography to zero when restarting a physgrid run.
 
This also fixes an issue running certain physgrid configurations, such as ne4pg2, because the default NTASK was set higher than the number of physics columns. To get the physgrid specification to match correctly in config_pesall.xml an underscore was to be added to other grid query strings. This should not affect non-physgrid cases. The underscore is used because it always separates the atmosphere grid from the other grids. For example, A typical ne4 grid string is
a%ne4np4_l%ne4np4_oi%ne4np4_r%null_g%null_w%null_z%null_m%oQU240
But the string for ne4pg2 is
a%ne4np4.pg2_l%ne4np4_oi%ne4np4_r%null_g%null_w%null_z%null_m%oQU240
so in order to differentiate the grids we use "ne4np4_" and "ne4np4.pg2" to get different default PE layouts. Using "ne4np4" without the underscore leads to an error when running physgrid cases.

After some discussion we decided to also include a change to the default physgrid mapping algorithm to use the high-order option. This will make all pg2 tests non-BFB.

[BFB] except for SMS_Ln5.ne4pg2_ne4pg2.FC5AV1C-L

### CODE REVIEW

File: components/cam/src/dynamics/se/inidat.F90 [hunk 1]
```diff
@@ -475,6 +471,13 @@ subroutine read_inidat( ncid_ini, ncid_topo, dyn_in)
     end if
 
     if (fv_nphys == 0) then
+      ! Map phis data to FV physics grid
```
Comment (ambrad) [inline components/cam/src/dynamics/se/inidat.F90]: This appears not to be right. The conditional at line 473 is `fv_nphys == 0`, but lines 474 to 479 then perform physgrid operations.

File: components/cam/src/dynamics/se/inidat.F90 [hunk 2]
```diff
@@ -453,18 +453,14 @@ subroutine read_inidat( ncid_ini, ncid_topo, dyn_in)
 
     if ( (ideal_phys .or. aqua_planet)) then
        tmp(:,1,:) = 0._r8
+       phis_tmp(:,:) = 0._r8
     else    
       fieldname = 'PHIS'
       tmp(:,1,:) = 0.0_r8
       if (fv_nphys > 0) then
-         ! Copy phis field to GLL grid
+         ! Load phis field to GLL grid first
```
Comment (ambrad) [inline components/cam/src/dynamics/se/inidat.F90]: I believe this comment is incorrect when separated from the fv_phys_to_dyn_topo function calls. Here, infld is loading physgrid topo data, not GLL data.

File: components/cam/src/dynamics/se/inidat.F90 (hunk 1 above)
Comment (whannah1) [reply components/cam/src/dynamics/se/inidat.F90]: oh, shoot, good catch. That's a pretty dumb mistake.

File: components/cam/src/dynamics/se/inidat.F90 (hunk 2 above)
Comment (whannah1) [reply components/cam/src/dynamics/se/inidat.F90]: Right, I'll correct that.

**Comment by whannah1** [comment]
@AaronDonahue , I couldn't find any other instances of phys_tmp

**Comment by oksanaguba** [comment]
@whannah1 I see some recent commits, are you still working on this?

**Comment by whannah1** [comment]
@oksanaguba, I noticed a couple minor things that don't seem to have any effect (when testing on Cori at least). I think this branch is still good to go, but I was gonna ask @AaronDonahue to give it a quick look over again. I just need another day or two to verify these fixes also work for the ECP fork, because there's a (perhaps different?) restart issue there as well that I'm struggling to understand.

**Comment by whannah1** [comment]
Ok, so the problem that I thought I had appears to be a non-issue. @AaronDonahue, can you take a quick look at the recent changes? Afterwards this PR should be ready to go.

**Comment by rljacob** [comment]
@AaronDonahue this needs another review from you.

**Comment by mt5555** [comment]
2nding @rljacob request for another review from @AaronDonahue !

**Comment by AaronDonahue** [comment]
My apologies, I've been in and out of travel and missed this pinging of me.  I'll get on this right away.

**Comment by oksanaguba** [comment]
@whannah1 did you run any tests? description says it is fixing restart pg tests, did you run them with these fixes? if so, please, write down a comment saying which tests you ran on which machine (command line?).

**Comment by whannah1** [comment]
@oksanaguba I generally don't use the testing framework unless I'm merging PR because I'm often interested in looking at model output. I should get in the habit of running tests more, but it's so difficult to keep track of everything when you have wait hours in between tests and I'm always jumping back and forth between branches and repos. I know some people have better ways of handling this and I'm open to any suggestions you might have. 

I'll start some pg2 and np4 tests now.

Good catch on the "topo" variable. It isn't used, so I'll delete it.

**Comment by whannah1** [comment]
I ran this series of tests (on Cori), which all passed
  SMS_Ln5.ne4pg2_ne4pg2.FC5AV1C-L.cori-knl_intel
  SMS_Ln5.ne4_ne4.FC5AV1C-L.cori-knl_intel
  SMS.ne4_ne4.FC5AV1C-L.cori-knl_intel.cam-thetahy_ftype2
  SMS_Ln5.ne4pg2_ne4pg2.FC5AV1C-L.cori-knl_intel.cam-thetahy_pg2

**Comment by oksanaguba** [comment]
Thanks. You made baselines, right? It seems that there are no pg2 restart tests. Should we add one? (not necessarily in this PR).

**Comment by whannah1** [comment]
I did not make baselines. I can do a restart test. 
Are you suggesting we add an aqua restart test or a normal restart test?

**Comment by oksanaguba** [comment]
Please, rerun your tests against 'baselines' as below.

Here is my script that I use for anvil to run tests. I do not use machines' baselines even if they exist, easiest is to generate your own with next branch that is at the same state as your merged PR. That is, if you suspect that your PR may diverge too much from master and next, then pull next into 1 clone. Pull next into another clone, merge your branch into the second clone, and then run these tests. First clone will be used to generate baselines. Uncomment/modify for your tests and machine.

```
#!/bin/bash -e

brr=~/acme-b2
master=~/acme-next ##### NEXT!!!
baselinefold=~/as/acme-tests

#test=e3sm_developer
test=HOMME_P24.f19_g16_rx1.A.anvil_intel

#generate baseline from master
${master}/cime/scripts/create_test -g -o --baseline-root ${baselinefold} -b master ${test}
#run test from branch
${brr}/cime/scripts/create_test -c --baseline-root ${baselinefold} -b master ${test}

test=ERP_Ln9.ne4_ne4.FC5AV1C-L.anvil_intel

#generate baseline from master
#${master}/cime/scripts/create_test -g -o --baseline-root ${baselinefold} -b master ${test}
#run test from branch
#${brr}/cime/scripts/create_test -c --baseline-root ${baselinefold} -b master ${test}

test=eam_theta
suff=' --machine anvil --compiler intel '

#generate baseline from master
${master}/cime/scripts/create_test $suff -g -o --baseline-root ${baselinefold} -b master ${test}
#run test from branch
${brr}/cime/scripts/create_test    $suff -c --baseline-root ${baselinefold} -b master ${test}

test=eam_preqx
suff=' --machine anvil --compiler intel '

#generate baseline from master
#${master}/cime/scripts/create_test $suff -g -o --baseline-root ${baselinefold} -b master ${test}
#run test from branch
#${brr}/cime/scripts/create_test    $suff -c --baseline-root ${baselinefold} -b master ${test}
```

**Comment by oksanaguba** [comment]
About restart: I do not know about aquaplaned, but usual ERS test for pg2 would be good. We probably should start grouping pg2 tests into a group that shares the same exec. This can be a separate PR, probably, easier, because it will be nonbfb.

**Comment by whannah1** [comment]
This PR is fixing the restart capability for pg2 aqua runs only, so adding a non-aqua restart for pg2 would not have caught this. So if we want full test coverage we should add both aqua and non-aqua

**Comment by oksanaguba** [comment]
Ok, let's do it in a separate PR.

File: components/cam/src/dynamics/se/dyn_grid.F90 [hunk 3]
```diff
@@ -84,6 +84,7 @@ module dyn_grid
     real(kind=r8), allocatable :: lat(:,:)    ! latitude
     real(kind=r8), allocatable :: corner_lon(:,:,:)    ! longitude
     real(kind=r8), allocatable :: corner_lat(:,:,:)    ! latitude
+    real(kind=r8), allocatable :: topo(:,:)   ! topography data
```
[the remark this answers is not in the document]
Comment (whannah1) [reply components/cam/src/dynamics/se/dyn_grid.F90]: yes, it is used in fv_physics_coupling_mod.F90

**Comment by ambrad** [comment]
@whannah1, in case this is of any use, @oksanaguba helped me set up regression testing of my PRs a while ago. I took her script and edited it a bit to suit my style. This perhaps simpler script may be of use to you.

The idea is to have one repo, I call mine e3sm1, for regression testing only. (It's not for dev b/c I want to keep working on various branches.) When I want to test a PR, I bring master and my PR branch up to date in e3sm1. Then I run the following script:
```
#!/bin/bash -e

e3sm=/ascldap/users/ambradl/climate/e3sm1 # up-to-date repo used just for regression testing
#wcid=   # fill in the project ID

branch=ambrad/eam/fix-3241  # name of branch to be tested

test=e3sm_developer
baselinefold=/gscratch/ambradl/e3smbaselines  # place to store baseline files

# generate baseline from master
(cd $e3sm; git checkout master; git submodule update; git log --oneline -n 2 > tmp.txt; cat tmp.txt)
${e3sm}/cime/scripts/create_test --wait -p $wcid -g -o --baseline-root=${baselinefold} -r ${baselinefold}/testroot -b master ${test}

# run test from branch
(cd $e3sm; git checkout $branch; git submodule update; git log --oneline -n 2 > tmp.txt; cat tmp.txt)
${e3sm}/cime/scripts/create_test --wait -p $wcid -c --baseline-root=${baselinefold} -r ${baselinefold}/testroot -b master ${test}
```

**Comment by whannah1** [comment]
@ambrad I have multiple scripts for doing this already, and I understand the basic process. My issues are with organization and the creating a suitable baseline. This PR is over a month old now, so I can't just used the current master, it seems more appropriate to jump to a detached head state to get the exact version of master that I started with. The alternative is to merge with current master and hope there aren't conflicts that might further obscure the attribution of diffs if they happen. 

So when you say you "bring master and my PR branch up to date in e3sm1" you mean you rebase on the current master? Or merge with master in a 3rd branch?

**Comment by ambrad** [comment]
Agreed. That's why I don't include updating master in the branch as an automatic thing. I do one of two things, depending on what I want:
1. Manually set up master in my e3sm1 repo to be what I want, using a hard reset to the SHA1 I branched from.
2. Or merge master into my branch and run against the current E3SM-Project/master.
Then I run the script. I go with 2 if there are no conflicts with current master, 1 if there are. But to be clear, the PR that will be merged does not have the merged commits from master; I do that just in my PR-testing repo. Obviously one can think of many variations on these two options.

**Comment by whannah1** [comment]
So there's a new problem here that has to do with the low-order mapping option being the default. The pg2 runs are failing the restart test because the topo field goes through an extra smoothing step on restart, which leads to a diff. It seems wasteful to fix this because we've already decided that we want to completely remove the low-order option. 

It's a bit hard to decide what to do here, but I think the best course of action is to include the change in default mapping algorithm from PR #3251 in this current PR. This will resolve the issue because the high-order option already maps back and forth between GLL and FV grids to avoid the restart problem. After this and PR #3267 get merged, then I'll do a new PR to remove the low-order option, which should be straightforward.

**Comment by oksanaguba** [comment]
@whannah1 if this is not affecting nightlies, I would think it is ok to leave as is. This is my understanding that there are no restart pg2 tests, which would be a next PR, restart tests with aquaplanet and without.

**Comment by whannah1** [comment]
Over the weekend I ran these tests on the updated branch, which all passed comparisons with baselines.
- SMS_Ln5.ne4_ne4.FC5AV1C-L.cori-knl_intel
- ERS_Ld3.ne4_ne4.FC5AV1C-L.cori-knl_intel
- ERP_Ld3.ne4_ne4.FC5AV1C-L.cori-knl_intel

These pg2 tests also passed (no baseline comparison due to change in the mapping algorithm)
- ERP_Ld3.ne4pg2_ne4pg2.FC5AV1C-L.cori-knl_intel
- ERS_Ld3.ne4pg2_ne4pg2.F-EAMv1-AQP1.cori-knl_intel

**Comment by oksanaguba** [comment]
Thanks @whannah1 . I will merge it this week.

**Comment by oksanaguba** [comment]
@whannah1 is PR supposed to be bfb? I am getting   SMS_Ln5.ne4pg2_ne4pg2.FC5AV1C-L fail against master as of today (compared to the same master with this PR merged).

**Comment by oksanaguba** [comment]
I assume these changes won't make pg2 runs bfb?
``` 
-   se_fv_phys_remap_alg = 0
+     se_fv_phys_remap_alg = 1
```

**Comment by whannah1** [comment]
@oksanaguba The pg2 runs are not expected to be BFB as of the latest commit that changes the default mapping algorithm. I forgot to update the description, but I'll do that now.

**Comment by oksanaguba** [comment]
I see a lot of     

FAIL ERS.ne4_ne4.FC5AV1C-L.anvil-centos7_intel.cam-thetanh_ftype2 MEMCOMP Error: Memory usage increase > 10% from baseline 

from running e3sm_atm_integration .

**Comment by oksanaguba** [comment]
I see 1 new allocate and new calls to PG remap routines. 10% sounds like extreme for just 1 allocate of topo. @ambrad @whannah1 any ideas?

**Comment by ambrad** [comment]
First, @oksanaguba, just to be clear, the increase is happening in a non-physgrid case, right? @whannah1, if you added allocs, you should make sure they run only when physgrid is requested.

**Comment by oksanaguba** [comment]
Yes, but even with that 10% is too much?

**Comment by whannah1** [comment]
@oksanaguba, that's very strange. I don't see how that could happen.

**Comment by oksanaguba** [comment]
There are also fails like this `FAIL SMS.ne4_ne4.FC5AV1C-L.anvil-centos7_intel.cam-preqx_ftype4 TPUTCOMP Error: Computation time increase > 10 pct from baseline`
but I usually ignore these, tests are too short.

**Comment by ambrad** [comment]
phis_tmp is not deallocated in inidat.F90, which should be cleaned up, but I believe an allocatable is supposed to dealloc automatically.

**Comment by whannah1** [comment]
fv_physgrid_init is only called when fv_nphys>0 in dyn_comp.F90, so that can't be causing the memory usage change

**Comment by oksanaguba** [comment]
mem check may not be perfect, but whom would we ask?

**Comment by whannah1** [comment]
I can add that deallocate statement

**Comment by whannah1** [comment]
@oksanaguba , what's the run length of this test?
ERS.ne4_ne4.FC5AV1C-L.anvil-centos7_intel.cam-thetanh_ftype2

**Comment by oksanaguba** [comment]
memory output, master and then PR:
```
(base) [onguba@blueslogin4 run]$ gunzip cpl.log.88244.191105-170925.gz
(base) [onguba@blueslogin4 run]$ grep highwater cpl.log.88244.191105-170925
 memory_write: model date =   00010102       0 memory =  156081.60 MB (highwater)      58845.44 MB (usage)  (pe=    0 comps= cpl ATM ICE GLC WAV IAC ESP)
 memory_write: model date =   00010103       0 memory =  156081.92 MB (highwater)      59142.72 MB (usage)  (pe=    0 comps= cpl ATM ICE GLC WAV IAC ESP)
 memory_write: model date =   00010104       0 memory =  156081.92 MB (highwater)      59145.92 MB (usage)  (pe=    0 comps= cpl ATM ICE GLC WAV IAC ESP)
 memory_write: model date =   00010105       0 memory =  156081.92 MB (highwater)      59150.08 MB (usage)  (pe=    0 comps= cpl ATM ICE GLC WAV IAC ESP)
 memory_write: model date =   00010106       0 memory =  156081.92 MB (highwater)      59324.48 MB (usage)  (pe=    0 comps= cpl ATM ICE GLC WAV IAC ESP)
(seq_mct_drv): ===============  pes min memory highwater  (MB)    1120.967  ===============
(seq_mct_drv): ===============  pes max memory highwater  (MB)  465381.000  ===============
(base) [onguba@blueslogin4 run]$
(base) [onguba@blueslogin4 run]$ cd ../../SMS.ne4_ne4.FC5AV1C-L.anvil-centos7_intel.cam-thetanh_ftype2.C.20191105_172101_ewo6rl/run/
(base) [onguba@blueslogin4 run]$ gunzip cpl.log.88278.191105-174828.gz
(base) [onguba@blueslogin4 run]$ grep highwater cpl.log.88278.191105-174828
 memory_write: model date =   00010102       0 memory =  185810.67 MB (highwater)      71104.00 MB (usage)  (pe=    0 comps= cpl ATM ICE GLC WAV IAC ESP)
 memory_write: model date =   00010103       0 memory =  185811.05 MB (highwater)      71373.33 MB (usage)  (pe=    0 comps= cpl ATM ICE GLC WAV IAC ESP)
 memory_write: model date =   00010104       0 memory =  185811.05 MB (highwater)      71435.81 MB (usage)  (pe=    0 comps= cpl ATM ICE GLC WAV IAC ESP)
 memory_write: model date =   00010105       0 memory =  185811.05 MB (highwater)      71438.48 MB (usage)  (pe=    0 comps= cpl ATM ICE GLC WAV IAC ESP)
 memory_write: model date =   00010106       0 memory =  185811.05 MB (highwater)      71691.05 MB (usage)  (pe=    0 comps= cpl ATM ICE GLC WAV IAC ESP)
(seq_mct_drv): ===============  pes min memory highwater  (MB)    1109.342  ===============
(seq_mct_drv): ===============  pes max memory highwater  (MB)  248205.867  ===============
```

**Comment by ambrad** [comment]
Are the number of processes the same in master and PR runs? Maybe the PE XML changes are somehow (erroneously) changing that, giving fewer processes with consequently more memory per process.

**Comment by oksanaguba** [comment]
Some sanity checks: both clones are from yesterday's master.

Ranks: Both have global communicator with 144 ranks, both homme outputs have 96 ranks, 1 element per rank.

**Comment by oksanaguba** [comment]
@whannah1 tests ERS.ne4_ne4.FC5AV1C-L.anvil-centos7_intel.cam-thetanh_ftype2 ran for < 2 minutes.

**Comment by whannah1** [comment]
@oksana I meant the run length, since it's not specified in the test name. What's the default run length for an ERS test?

**Comment by oksanaguba** [comment]
11 days for that ERS test.

**Comment by whannah1** [comment]
This might be a dumb question, but could a change in the submodules explain the change in memory use? Maybe one of the clones didn't sync those properly?

**Comment by oksanaguba** [comment]
Both runs were done on anvil in the same terminal window. But I will resubmit PR run now. I also looked at some timers and all looks ok.

**Comment by oksanaguba** [comment]
I misread your question. submodules. I will resync clones.

**Comment by oksanaguba** [comment]
Baselines clone had older commits in submodules, tests are running now.

**Comment by whannah1** [comment]
ok, let's hope that's the explanation.

**Comment by oksanaguba** [comment]
Here are two runs from master, (first) yesterday with older commits for app, mpas, and scorpio, and today's the same clone with synced modules (second):
```
(base) [onguba@blueslogin2 run]$ cd ../../SMS.ne4_ne4.FC5AV1C-L.anvil-centos7_intel.cam-thetanh_ftype2.G.20191105_164323_gq0oc1/run/
(base) [onguba@blueslogin2 run]$ grep highwater cpl.log.*
 memory_write: model date =   00010102       0 memory =  156081.60 MB (highwater)      58845.44 MB (usage)  (pe=    0 comps= cpl ATM ICE GLC WAV IAC ESP)
 memory_write: model date =   00010103       0 memory =  156081.92 MB (highwater)      59142.72 MB (usage)  (pe=    0 comps= cpl ATM ICE GLC WAV IAC ESP)
 memory_write: model date =   00010104       0 memory =  156081.92 MB (highwater)      59145.92 MB (usage)  (pe=    0 comps= cpl ATM ICE GLC WAV IAC ESP)
 memory_write: model date =   00010105       0 memory =  156081.92 MB (highwater)      59150.08 MB (usage)  (pe=    0 comps= cpl ATM ICE GLC WAV IAC ESP)
 memory_write: model date =   00010106       0 memory =  156081.92 MB (highwater)      59324.48 MB (usage)  (pe=    0 comps= cpl ATM ICE GLC WAV IAC ESP)
(seq_mct_drv): ===============  pes min memory highwater  (MB)    1120.967  ===============
(seq_mct_drv): ===============  pes max memory highwater  (MB)  465381.000  ===============
(base) [onguba@blueslogin2 run]$ cd ../../SMS.ne4_ne4.FC5AV1C-L.anvil-centos7_intel.cam-thetanh_ftype2.G.20191106_101124_ih5vi9/run/
(base) [onguba@blueslogin2 run]$ grep highwater cpl.log.*
 memory_write: model date =   00010102       0 memory =  975524.00 MB (highwater)     370904.00 MB (usage)  (pe=    0 comps= cpl ATM ICE GLC WAV IAC ESP)
 memory_write: model date =   00010103       0 memory =  975526.00 MB (highwater)     372420.00 MB (usage)  (pe=    0 comps= cpl ATM ICE GLC WAV IAC ESP)
 memory_write: model date =   00010104       0 memory =  975526.00 MB (highwater)     372646.00 MB (usage)  (pe=    0 comps= cpl ATM ICE GLC WAV IAC ESP)
 memory_write: model date =   00010105       0 memory =  975526.00 MB (highwater)     372672.00 MB (usage)  (pe=    0 comps= cpl ATM ICE GLC WAV IAC ESP)
 memory_write: model date =   00010106       0 memory =  975526.00 MB (highwater)     373930.00 MB (usage)  (pe=    0 comps= cpl ATM ICE GLC WAV IAC ESP)
(seq_mct_drv): ===============  pes min memory highwater  (MB)    1111.784  ===============
(seq_mct_drv): ===============  pes max memory highwater  (MB) 1249578.667  ===============
```

**Comment by ambrad** [comment]
If I'm reading these data correctly, (1) the second has >2x larger "max memory highwater" and >5x highwater on rank 0 but (2) the only difference is submodule SHA1s. Do others agree? If I have those two points right, then there seems to be either noise or a real issue due to a submodule change but likely nothing related to this PR.

**Comment by oksanaguba** [comment]
yes, has nothing to do with this pr. I will make an issue.

**Comment by oksanaguba** [comment]
Reran tests, e3sm_atm_integration, against master with updated clone. all baselines pass, memory issues are not from this pr.

**Comment by whannah1** [comment]
that's good, thanks for all the extra effort!

**Comment by oksanaguba** [comment]
merged to next

**Comment by oksanaguba** [comment]
melvin does not allow 96 cores for pg2 runs, limit is 60. a fix was pushed, I need to wait for another day.

**Comment by oksanaguba** [comment]
remerged to next

**Pull request merged**